plugins/Validation/src/Validation.ts

```diff
@@ -1,223 +1,346 @@
-                    result = result.concat(
-                        Validation(nestedInst[k] as State<StateValueAtPath>)
-                            .errors(filter, depth - 1, first));
-                    if (first && result.length > 0) {
-                        return result;
-                    }
+            target = nested.parent;
+
+            while (target.state.path.filter(f => typeof f !== 'number').toString() !== condition.path.toString()) {
+                if (target.parent) {
+                    target = target.parent;
+                } else {
+                    throw new Error('not sure how to handle');
                 }
             }
+
+            target = target.state;
         }
-        return consistentResult();
+
+        return condition.fn(target, condition.root.state);
     }
 }
 
-// tslint:disable-next-line: function-name
-export function Validation(): Plugin;
-export function Validation<S>($this: State<S>): ValidationExtensions<S>;
-export function Validation<S>($this?: State<S>): Plugin | ValidationExtensions<S> {
-    if ($this) {
-        let state = $this;
+function isPrimitiveState<T>(state: any): state is State<T> {
+    return state.keys === undefined;
+}
 
-        const [plugin] = state.attach(PluginID);
-        if (plugin instanceof Error) {
-            throw plugin
-        }
-        const instance = plugin as ValidationPluginInstance<S>;
-
-        const inst = instance;
-        return {
-            validate: (r, m, s) => {
-                inst.addRule(state.path, {
-                    rule: r,
-                    message: m,
-                    severity: s || 'error'
-                })
-            },
-            validShallow(): boolean {
-                return inst.getErrors(state, 1, undefined, true).length === 0
-            },
-            valid(): boolean {
-                return inst.getErrors(state, Number.MAX_SAFE_INTEGER, undefined, true).length === 0
-            },
-            invalidShallow(): boolean {
-                return inst.getErrors(state, 1, undefined, true).length !== 0
-            },
-            invalid(): boolean {
-                return inst.getErrors(state, Number.MAX_SAFE_INTEGER, undefined, true).length !== 0
-            },
-            errors: (filter, depth, first) => {
-                return inst.getErrors(state, depth === undefined ? Number.MAX_SAFE_INTEGER : depth, filter, first);
-            },
-            firstError: (filter, depth) => {
-                const r = inst.getErrors(state, depth === undefined ? Number.MAX_SAFE_INTEGER : depth, filter, true);
-                if (r.length === 0) {
-                    return {};
-                }
-                return r[0];
-            },
-        }
+function isObjectState<T>(state: State<T>, keys: any): keys is ReadonlyArray<keyof T> {
+    return Array.isArray(state.keys) && state.keys.length > 0 && state.keys.every(k => typeof k === 'string');
```

Comment:
Better to make type detection consistent with the core package:

If Array.isArray(state) is true, it is array. If `typeof state === 'object' && state !== null`, it is an object, otherwise it is primitive type.

Reply:
that is def better